File: auth-service/app/Models/Asset.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Asset extends Model
{
    use SoftDeletes;

    protected $fillable = [
        'cab_number',
        'project_code',
        'asset_number',
        'designation',
        'serial_number',
        'quantity',
        'unit_price',
        'delivery_date',
        'status',
        'service_start_date',
        'exit_date',
        'exit_reason',
        'supplier_id',
        'project_id',
        'asset_type_id',
        'location_id',
    ];

    /**
     * The attributes that should be cast.
     */
    protected $casts = [
        'delivery_date' => 'date',
        'service_start_date' => 'date',
        'exit_date' => 'date',
        'unit_price' => 'decimal:2',
        'quantity' => 'integer',
    ];

    /**
     * Get the supplier that owns the asset.
     */
    public function supplier(): BelongsTo
    {
        return $this->belongsTo(Supplier::class);
    }

    /**
     * Get the project that owns the asset.
     */
    public function project(): BelongsTo
    {
        return $this->belongsTo(Project::class);
    }

    /**
     * Get the asset type.
     */
    public function assetType(): BelongsTo
    {
        return $this->belongsTo(AssetType::class);
    }

    /**
     * Get the location of the asset.
     */
    public function location(): BelongsTo
    {
        return $this->belongsTo(Location::class);
    }

    /**
     * Get the status history of the asset.
     */
    public function statusHistories(): HasMany
    {
        return $this->hasMany(AssetStatusHistory::class)->orderBy('created_at', 'desc');
    }
}

File: inventory-service-spring/src/main/java/com/inventory/asset/entity/Asset.java
package com.inventory.asset.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "assets")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Asset {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "cab_number")
    private String cabNumber;

    @Column(name = "project_code")
    private String projectCode;

    @Column(name = "asset_number")
    private String assetNumber;

    private String designation;

    @Column(name = "serial_number")
    private String serialNumber;

    private Integer quantity;

    @Column(name = "unit_price", precision = 15, scale = 2)
    private BigDecimal unitPrice;

    @Column(name = "delivery_date")
    private LocalDate deliveryDate;

    private String status;

    @Column(name = "service_start_date")
    private LocalDate serviceStartDate;

    @Column(name = "exit_date")
    private LocalDate exitDate;

    @Column(name = "exit_reason")
    private String exitReason;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "supplier_id")
    private Supplier supplier;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_id")
    private Project project;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "asset_type_id")
    private AssetType assetType;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "location_id")
    private Location location;

    @OneToMany(mappedBy = "asset", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("createdAt DESC")
    @Builder.Default
    private List<AssetStatusHistory> statusHistories = new ArrayList<>();

    @Column(name = "deleted_at")
    private LocalDateTime deletedAt;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public void addStatusHistory(AssetStatusHistory history) {
        statusHistories.add(history);
        history.setAsset(this);
    }
}

